src/test/shared/codelens/localLambdaRunner.test.ts

```diff
@@ -0,0 +1,323 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+import * as assert from 'assert'
+import * as vscode from 'vscode'
+import { DebugConfiguration } from '../../../lambda/local/debugConfiguration'
+import * as localLambdaRunner from '../../../shared/codelens/localLambdaRunner'
+import { BasicLogger, ErrorOrString } from '../../../shared/logger'
+import { ChannelLogger } from '../../../shared/utilities/vsCodeUtils'
+
+class FakeChannelLogger implements Pick<ChannelLogger, 'info' | 'error' | 'logger'> {
+    public readonly loggedInfoKeys: Set<string> = new Set<string>()
+    public readonly loggedErrorKeys: Set<string> = new Set<string>()
+    public readonly logger: FakeBasicLogger = new FakeBasicLogger()
+
+    public info(nlsKey: string, nlsTemplate: string, ...templateTokens: ErrorOrString[]): void {
+        this.loggedInfoKeys.add(nlsKey)
+    }
+
+    public error(nlsKey: string, nlsTemplate: string, ...templateTokens: ErrorOrString[]): void {
+        this.loggedErrorKeys.add(nlsKey)
+    }
+}
+
+class FakeBasicLogger implements BasicLogger {
+    public readonly loggedDebugEntries: ErrorOrString[] = []
+
+    public debug(...message: ErrorOrString[]): void {
+        this.loggedDebugEntries.push(...message)
+    }
+
+    public verbose(...message: ErrorOrString[]): void {
+        throw new Error('verbose() not used')
+    }
+
+    public info(...message: ErrorOrString[]): void {
+        throw new Error('info() not used')
+    }
+
+    public warn(...message: ErrorOrString[]): void {
+        throw new Error('warn() not used')
+    }
+
+    public error(...message: ErrorOrString[]): void {
+        throw new Error('error() not used')
+    }
+}
+
+describe('LocalLambdaRunner', async () => {
+    describe('attachDebugger', async () => {
+        let actualRetries: number = 0
+        let channelLogger: FakeChannelLogger
+
+        beforeEach(async () => {
+            actualRetries = 0
+            channelLogger = new FakeChannelLogger()
+        })
+
+        async function onWillRetry(): Promise<void> {
+            actualRetries++
+        }
+
+        async function startDebuggingReturnsTrue(
+            folder: vscode.WorkspaceFolder | undefined,
+            nameOrConfiguration: string | vscode.DebugConfiguration
+        ): Promise<boolean> {
+            return Promise.resolve(true)
+        }
+
+        async function startDebuggingReturnsFalse(
+            folder: vscode.WorkspaceFolder | undefined,
+            nameOrConfiguration: string | vscode.DebugConfiguration
+        ): Promise<boolean> {
+            return Promise.resolve(false)
+        }
+
+        async function startDebuggingReturnsUndefined(
+            folder: vscode.WorkspaceFolder | undefined,
+            nameOrConfiguration: string | vscode.DebugConfiguration
+        ): Promise<boolean> {
+            const result: boolean | undefined = undefined
+
+            return Promise.resolve(result!)
```

Comment:
Correct me if I'm wrong, but this shouldn't return `undefined`, right?

Reply:
It does return undefined. This side-steps the linter calling it out.
We're emulating the `vscode.debugger.startDebugging` behavior which has the same signature, and returns `undefined`.